src/java.base/share/classes/java/lang/String.java

```diff
@@ -874,21 +866,34 @@ private static Charset lookupCharset(String csn) throws UnsupportedEncodingExcep
 
     private static byte[] encode(Charset cs, byte coder, byte[] val) {
         if (cs == UTF_8.INSTANCE) {
-            return encodeUTF8(coder, val, true);
+            return encodeUTF8(coder, val);
         }
         if (cs == ISO_8859_1.INSTANCE) {
             return encode8859_1(coder, val);
         }
         if (cs == US_ASCII.INSTANCE) {
             return encodeASCII(coder, val);
         }
-        return encodeWithEncoder(cs, coder, val, true);
+        return encodeWithEncoder(cs, coder, val);
+    }
+
+    private static byte[] encodeWithEncoder(Charset cs, byte coder, byte[] val) {
+        return encodeWithEncoder(cs, coder, val, null);
+    }
+
+    private static byte[] encodeWithEncoderNoReplacement(Charset cs, byte coder, byte[] val)
+            throws CharacterCodingException {
+        return encodeWithEncoder(cs, coder, val, CharacterCodingException.class);
     }
 
-    private static byte[] encodeWithEncoder(Charset cs, byte coder, byte[] val, boolean doReplace) {
+    private static <E extends Exception> byte[] encodeWithEncoder(
+            Charset cs, byte coder, byte[] val, Class<E> exceptionClass)
+            // Parametrizing on exception type to enable callers (using null) to avoid having to declare the exception
+            throws E {
```

Comment:
@RogerRiggs, since there is nothing stopping the programmer from passing `TotallyUnrelatedException.class` to `encodeWithEncoder()` (or to 2 other sneakily throwing methods), I've tried adding the following as the first thing in this methods' body:

    assert exceptionClass == null || CharacterCodingException.class.isAssignableFrom(exceptionClass);

This beautifully caused a JVM crash due to early initialization issues surrounding `String`. @liach suggested the following:

1. Move assertions to `StringLatin1` and `StringUTF16` – This solves the issue, but the check is at a relatively later stage.
2. Replace `assert` with a runtime check, e.g., `if (...) { throw new AssertionError(...); }` – This solves the issue, but incurs a runtime penalty

What would you advice?

Reply:
"the programmer" is an OpenJDK developer and won't be making spurious changes.
The exception classes are only present to make the compiler satisfied, they are ignored at the point the exceptions are thrown.
A mismatch in the exception class would show up at compile time, not matching the signature of the calling method.
And they are buried in the implementation under the `decodeUTF8_UTF16` and `decodeUTF8_UTF16NoReplacement` wrappers.